### ISSUE TITLE
Suggestion: invalidate cache for composer update vendor/package

### ISSUE THREAD

**Issue by mindplay-dk** [composer/composer#5292]
There is (still) a substantial delay between pushing a package update until you can actually install that update with `composer update`.

This is a long-standing issue for me and my coworkers.

For e.g. `composer update foo/bar`, how about invalidating the cached version of `foo/bar` so we don't have to wait?

I believe the issue is the local cache, as often I can see the updated package on packagist.org, but I can't get it until after a minute or two...

**Comment by stof** [comment]
The issue is not the local cache. Packagist is not generating the metadata files loaded by composer on the fly (this would be far too slow given the amount of packages on Packagist). Instead, a cron is dumping them to static files regularly (I don't know what is the current value of the interval, we will need to ask @Seldaek for that).
This is why there is a small delay between seeing it in the packagist interface (which happens as soon as the DB is updated) and seeing it in the composer client (which happens after the next run of the dumper)

**Comment by Seldaek** [comment]
The dumper runs once a minute so it really shouldn't be a very long wait on
average. It would be possible to improve this further but not without some
amount of work. Not saying it won't happen but it's not the top priority
right now.

**Comment by mindplay-dk** [comment]
Could this be improved by applying event sourcing?

If every change to the net state of the package registry was represented by ordered event objects, we wouldn't need to batch operations or introduce any arbitrary delay - a client would simply say, "I have records up to #12345" and the server would send any new events added after that; each client would apply the events, in order, to bring their local database up to current.

I know this is not something that could be implemented overnight, but do you think the idea is feasible?

**Comment by Seldaek** [comment]
I think we can greatly improve things without such a massive paradigm shift. I know event sourcing is the shit these days but please no.

**Comment by stof** [comment]
@mindplay-dk the local client never downloads the full Packagist database

**Comment by mindplay-dk** [comment]
I would never apply event sourcing (or any other pattern) because it's "the shit" - I would only apply it if it solves a real problem.

This problem is all too real. I'm working on a project right now where the dependency projects gets some 20 or so updates per day - waiting 1-2 minutes every time adds up and standing around tapping my feet causes me to lose focus and *kills* my productivity.

And this is a big project, so running `composer update` is time consuming to begin with.

If event sourcing is not the answer, what *can* we do? Any ideas?

I want to help, but I have no other ideas.

**Comment by stof** [comment]
@mindplay-dk having to replicate the whole Packagist database locally each time you run a ``composer update`` would not make it faster. Loading the whole package metadata in memory would require several tens GB of RAM, which is just not acceptable (it may already reach 1GB today when loading metadata only for packages related to your project) and would force to download tons of data.
And it would also make the Composer code much more complex, because it would then have to perform a projection of the event stream to rebuild the data it really needs (i.e. the current metadata).

**Comment by curry684** [comment]
> I'm working on a project right now where the dependency projects gets some 20 or so updates per day - waiting 1-2 minutes every time adds up and standing around tapping my feet causes me to lose focus and kills my productivity.

Isn't the real issue then that you are using the wrong workflow? All package managers, including the big ones like apt and yum, have implicitly delays on getting new packages available due to synchronization, reindexing etc. happening. As Packagist isn't as distributed (yet) it's actually quite fast still as it only has the 'reindexing' part happening.

If you are using projects for which it's vital you're using bleeding edge packages, every time all the time, you should install them once with `--prefer-source` and use Git pull afterwards to update - that's as instant as it gets. If the packages are developed in house you could also use path repositories to NFS shares and eliminate every manual intervention.

**Comment by Seldaek** [comment]
Closing as it's a packagist issue and well, kinda known problem that might be worked on at some point.

**Comment by mindplay-dk** [comment]
> Isn't the real issue then that you are using the wrong workflow?

More like the opposite? I think I'm using the correct workflow: I don't edit dependencies directly in the vendor folder - instead, if I find a bug or missing feature in a vendor package, I open that project, write regression-tests, add features etc., tag a release, then return to the first package and update.

**Comment by curry684** [comment]
> I think I'm using the correct workflow

You're relying on an external party's services for the speed of your workflow. Composer's ecosphere provides a multitude of options for not having to rely on these external dependencies, like most obviously path repositories or a self-hosted Satis repository. Both are perfect solutions to the problem as they have literally zero delay when set up correctly and operate on a glass-plate model that any package they cannot provide is still fetched from Packagist.

Hence my comment - if the speed of a third party's services is not to your liking, and there are several fine and simple alternatives for not having to depend on its speed, why not adapt the workflow?

**Comment by mindplay-dk** [comment]
> You're relying on an external party's services for the speed of your workflow

Actually, we used Satis up until about a year ago - it was slow and required too much maintenance, so we eventually switched to a self-hosted instance of Packagist. It's great in terms of maintenance, but suffers from the same delay problem as official Packagist of course.

Also, a growing number of our packages are open source, and eventually all of them will be, so we will need to use Packagist for all packages sooner or later.

Regardless of possible workarounds, it would be nice to have a solution to this problem that works for the entire community.

**Comment by curry684** [comment]
Out of curiosity: how did Satis become slow and high on maintenance? At my company we're also using Satis, configured to update automatically using Gitlab webhooks, and it's been literally zero maintenance and lightning fast for about a year now.

**Comment by mindplay-dk** [comment]
> Out of curiosity: how did Satis become slow and high on maintenance?

I think Gitlab manages the configuration files for you? So that takes away the maintenance burden. We use Bitbucket (then Stash) and at least back then, I don't think there was any integration for Satis.

As I recall, it also became slower and slower to update with every added package - we had a lot of packages, an update took several minutes, e.g. about as bad as it is now with our self-hosted Packagist, but we also switched for other reasons: Packagist has a front-end from which developers can browse and add new packages, rather than editing a big, messy JSON file.

I don't know the state of Satis today, I'm just telling you our situation back then, it may be a lot better now. On a side-note, I have been pushing for a switch from Bitbucket to Gitlab myself, but I'm afraid the organization is pretty much married to JIRA. (ugh.)

**Comment by mindplay-dk** [comment]
Er, that said, as mentioned, everything we build will eventually be open-source and listed on Packagist, which will then become our primary package repository - so, even if we did switch to Gitlab and Satis, we'll have the same problem again when we ultimately switch back to the regular public Packagist.

The only remaining argument for Satis at that point is as a cache for third-party packages, so we're not affected by Packagist downtime, but that's a pretty theoretical benefit as it's almost always available - and it further increases the time and burden of maintaining Satis, e.g. an additional buffer/layer further adding to the time it takes to update third-party packages, right?

I think, ultimately, it would be best for everyone if Packagist was quicker to update.
